pkg/istio-agent/xds_proxy_delta.go

```diff
@@ -188,36 +195,37 @@ func (p *XdsProxy) handleUpstreamDeltaResponse(con *ProxyConnection) {
 			// TODO: separate upstream response handling from requests sending, which are both time costly
 			proxyLog.Debugf("response for type url %s", resp.TypeUrl)
 			metrics.XdsProxyResponses.Increment()
-			// TODO(howardjohn) implement handlers. We need to make handlers take in a list of
-			// resources or something probably
-			//if h, f := p.handlers[resp.TypeUrl]; f {
-			//	err := h(resp)
-			//	var errorResp *google_rpc.Status
-			//	if err != nil {
-			//		errorResp = &google_rpc.Status{
-			//			Code:    int32(codes.Internal),
-			//			Message: err.Error(),
-			//		}
-			//	}
-			//	// Send ACK/NACK
-			//	con.requestsChan <- &discovery.DiscoveryRequest{
-			//		VersionInfo:   resp.VersionInfo,
-			//		TypeUrl:       resp.TypeUrl,
-			//		ResponseNonce: resp.Nonce,
-			//		ErrorDetail:   errorResp,
-			//	}
-			//	continue
-			//}
+			if h, f := p.handlers[resp.TypeUrl]; f {
+				if len(resp.Resources) == 0 {
+					// Empty response, nothing to do
+					// This assumes internal types are always singleton
```

Comment:
whats an example of an internal type?

Reply:
all of the ones in handlers - NDS, PCDS, etc. The stuff that istio-agent process locally instead of forwarding